File: src/main/java/com/senderman/miniroulette/game/Game.java
package com.senderman.miniroulette.game;

import com.senderman.miniroulette.exception.TooLateException;
import com.senderman.miniroulette.exception.TooLittleCoinsException;
import com.senderman.miniroulette.game.bet.Bet;
import io.micronaut.core.annotation.Nullable;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class Game<ID> {

    public final static int MAX_WAIT_TIME_SECONDS = 30;

    private final ID id;
    private final Map<Long, Player> players; // players by id;
    private final AtomicInteger timer;
    private boolean isOpenForBets;
    @Nullable
    private Integer currentCell;

    public Game(ID id) {
        this.id = id;
        this.players = new HashMap<>();
        this.timer = new AtomicInteger(0);
        isOpenForBets = true;
        currentCell = null;
        runTimer();
    }

    public synchronized void addBet(Player player, Bet bet) throws TooLateException, TooLittleCoinsException {
        if (!isOpenForBets)
            throw new TooLateException();

        if (bet.getAmount() < 2)
            throw new TooLittleCoinsException();

        resetTimer();

        if (players.containsKey(player.getId()))
            players.get(player.getId()).addBet(bet);
        else {
            player.addBet(bet);
            players.put(player.getId(), player);
        }
    }

    private void runTimer() {
        ForkJoinPool.commonPool().execute(() -> {
            while (timer.get() < MAX_WAIT_TIME_SECONDS) {
                timer.incrementAndGet();
                try {
                    TimeUnit.SECONDS.sleep(1);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
            spin();
        });
    }

    protected synchronized void spin() {
        isOpenForBets = false;
        currentCell = ThreadLocalRandom.current().nextInt(0, 13);

        if (currentCell == 0)
            processZero();
        else
            processNonZero(currentCell);
    }

    protected void processNonZero(int cell) {
        for (var player : getPlayers()) {
            int delta = 0;
            for (var bet : player.getBets()) {
                if (bet.isWin(cell)) {
                    delta += bet.getPay();
                } else {
                    delta -= bet.getAmount();
                }
            }
            player.setDelta(delta);
        }
    }

    protected void processZero() {
        for (var player : getPlayers()) {
            int delta = 0;
            for (var bet : player.getBets()) {
                if (bet.isWin(0)) {
                    delta += bet.getPay();
                } else {
                    delta -= bet.getAmount() / 2;
                }
            }
            player.setDelta(delta);
        }
    }

    private void resetTimer() {
        timer.set(0);
    }

    public ID getId() {
        return id;
    }

    public Collection<Player> getPlayers() {
        return players.values();
    }

    @Nullable
    public synchronized Integer getCurrentCell() {
        return currentCell;
    }
}

File: src/main/kotlin/com/senderman/miniroulette/gameobjects/Game.kt
package com.senderman.miniroulette.gameobjects

import com.senderman.miniroulette.MainHandler
import com.senderman.miniroulette.Services
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.thread

class Game(private val handler: MainHandler, val chatId: Long) {

    private val maxTime = 30
    private var waitingForBets = true
    private var timer: AtomicInteger = AtomicInteger(0)
    private val players = HashMap<Int, Player>() // id-player
    private var currentCell = -1
    private val messagesToDelete = HashSet<Int>()

    fun addBet(userId: Int, name: String, text: String, messageId: Int) {
        messagesToDelete.add(messageId)
        if (!waitingForBets) {
            handler.sendMessage(chatId, "Слишком поздно!", messageId)
            return
        }

        val amount = try {
            text.trim().split("\\s+".toRegex())[0].toInt()
        } catch (e: NumberFormatException) {
            handler.sendMessage(chatId, "Неверный формат!")
            return
        }
        if (amount < 2) {
            handler.sendMessage(chatId, "Ставка должна быть не меньше 2 монет!", messageId)
            return
        }

        if (Services.db.getCoins(userId) < amount) {
            handler.sendMessage(chatId, "У вас недостаточно денег!", messageId)
            return
        }

        val target = text.trim().toLowerCase().replace("^\\d+\\s+".toRegex(), "")
        val bet = try {
            Bet.createBet(amount, target)
        } catch (e: InvalidBetCommandException) {
            handler.sendMessage(chatId, "Неверный формат!")
            return
        } catch (e: InvalidBetRangeException) {
            handler.sendMessage(chatId, "Неверный диапазон!")
            return
        }

        if (userId in players)
            players[userId]!!.bets.add(bet)
        else {
            val player = Player(userId, name)
            player.bets.add(bet)
            players[userId] = player
        }

        Services.db.takeCoins(userId, amount)
        timer.set(0)
        messagesToDelete.add(handler.sendMessage(chatId, "Ставка принята!", messageId).messageId)
    }

    fun runTimer() {
        handler.sendMessage(chatId, "\uD83C\uDFB0 Делайте ваши ставки\n\n$fieldString")
        thread {
            while (timer.get() < maxTime) {
                timer.incrementAndGet()
                Thread.sleep(1000)
            }
            handler.sendMessage(chatId, "❇️ Ставки кончились, ставок больше нет")
            spin()
        }
    }


    private fun spin() {
        waitingForBets = false
        currentCell = ThreadLocalRandom.current().nextInt(0, 13)
        if (currentCell == 0)
            processZero()
        else
            processNonZero()

        val logEntry: String = currentCell.toString() + when {
            currentCell == 0 -> "💚"
            currentCell.isEven() -> "\uD83D\uDDA4"
            else -> "❤"
        }
        val log = (Services.db.getLog(chatId) ?: "")
            .lines()
            .takeLast(9)
            .joinToString(separator = "\n", truncated = "")
            .trim()
        val finalLog = "$log\n$logEntry"
        Services.db.setLog(chatId, finalLog)

        messagesToDelete.forEach { handler.deleteMessage(chatId, it) }
        handler.removeGame(this)
    }

    private fun processZero() {
        val text = StringBuilder("\uD83C\uDFB2 Итоги:\n")
        text.append("\uD83D\uDC9A 0\n\n")
        for (player in players.values) {
            text.appendln("<b>${player.name}:</b>")
            var profit = 0
            var delta = 0

            for (bet in player.bets) {
                if (bet.isWin(currentCell)) {
                    profit = bet.pay + bet.amount
                    delta += bet.pay
                    text.appendln("\uD83D\uDE0E +${bet.pay} (${bet.amount} на ${bet.stringTarget})")
                } else {
                    profit += bet.amount / 2
                    delta -= bet.amount / 2
                    text.appendln("\uD83D\uDE14 -${bet.amount / 2} (${bet.amount} на ${bet.stringTarget})")
                }
            }

            Services.db.addCoins(player.id, profit)
            text.appendln(formatDelta(delta)).appendln()
        }
        handler.sendMessage(chatId, text.trim().toString())
    }

    private fun processNonZero() {
        val text = StringBuilder("\uD83C\uDFB2 Итоги:\n")
        val colorEmoji = if (currentCell.isEven()) "\uD83D\uDDA4" else "❤️"
        text.append("$colorEmoji $currentCell\n\n")
        for (player in players.values) {
            text.appendln("<b>${player.name}:</b>")
            var profit = 0
            var delta = 0

            for (bet in player.bets) {
                if (bet.isWin(currentCell)) {
                    profit += bet.amount + bet.pay
                    delta += bet.pay
                    text.appendln("\uD83D\uDE0E +${bet.pay} (${bet.amount} на ${bet.stringTarget})")
                } else {
                    delta -= bet.amount
                    text.appendln("\uD83D\uDE14 -${bet.amount} (${bet.amount} на ${bet.stringTarget})")
                }
            }

            Services.db.addCoins(player.id, profit)
            text.appendln(formatDelta(delta)).appendln()
        }
        handler.sendMessage(chatId, text.trim().toString())
    }

    private fun Int.isEven() = this % 2 == 0
    private fun formatDelta(delta: Int) = if (delta > 0) "\uD83D\uDCC8: +$delta" else "\uD83D\uDCC9: $delta"

    companion object {
        val fieldString = """
            0💚 
            1❤️ 2🖤 3❤️ 4🖤 5❤️ 6🖤
            7❤️ 8🖤 9❤️ 10🖤 11❤️ ️12🖤
        """.trimIndent()
    }

}